File: filter/filter.py
import metric_tree

class Filter:

    base_ban_list = ["apple", "orange", "flatiron"]
    base_allowed_list = ["the", "is", "a"]

    replacement_dict = { "@" : ("a", -.2), "!" : ("i", -.2), "&" : ("an", .2), "." : ("", -.05), "$" : ("s", -.2), "3" : ("e", -.2) }

    tree = metric_tree.Tree()
    tolerance = .85

    def __init__(self):
        for banned in self.base_ban_list:
            self.tree.insert_word(banned)

    def add_word(self, word):
        self.tree.insert_word(word)

    def censor_sentence(self, sentence, user_pcs=0):
        words = sentence.split(' ')
        new_pcs = user_pcs

        for word in words:
            modifier = 0
            check_word = word.lower()

            if word not in self.base_allowed_list:
                for i in range(len(check_word)):
                    if check_word[i] in self.replacement_dict:
                        ret_tup = self.replacement_dict[word[i]]
                        check_word = check_word[:i] + ret_tup[0] + check_word[i+1:]
                
                score = self.tree.get_score(check_word, modifier)

                if score > self.tolerance-(user_pcs/100):
                    sentence = sentence.replace(word, "*"*len(word))
                    new_pcs += 1
        
        return sentence, new_pcs


filt = Filter()

print(filt.censor_sentence("the @pple is orang3", 5)[0])

File: filter/filter.go
package filter

const baseConfidence = .75

type Filter struct {
	tree *MetricTree
	toleranceModifier float64
	wordSet *wordSet
}

/*  Creates a new filter.  That large list is the 1000 most common words in English.  This
 *  This greatly reduces execution time by providing a large bank of whitelisted terms.  The
 *  most expensive operation in the filter is scoring a word.  Even though it takes time to
 *  iterate over this list, and this line is ridiculously long, it improves execution time
 *  by a factor of 2 in some instances.
 */
func NewFilter(bannedTree *MetricTree) *Filter {
	return &Filter{
		tree: bannedTree,
		wordSet: newWordSet(),
	}
}

/* Inserts a word into the tree of banned words */
func (filter *Filter) BanWord(word string) {
	filter.tree.insertWord(word)
}


/*  This takes in the character and replaces it if needed.  Common special character substitutes
 *  are replaced (@ -> a).  If a change takes place, the modifier is updated and returned.
 */

/*  The general test scores the entire input word. */
func (filter *Filter) TestGeneral(input []rune) bool {
	pass := true
	testSlice, _ := filter.replaceSymbols(input)
	if !filter.wordSet.isKnown(testSlice) {
		score := filter.tree.getScore(string(testSlice), 0)
		if score > (baseConfidence + filter.toleranceModifier) {
			pass = false
		}
	}
	return pass
}

/*  The period test first removes all of the periods and commas.  Whenever a removal takes place,
 *  it assesses a large penalty to the modifier.  A rolling selection of four characters from the
 *  edited rune slice is then tested.  If any tested subset receives a high enough score, the entire
 *  unedited string is failed.
 */
func (filter *Filter) TestRolling(input []rune) bool {
	score := 0.0
	pass := true
	testSlice, modifier := filter.replaceSymbols(input)
	if !filter.wordSet.isKnown(testSlice) {
		if len(testSlice) > 4 {
			for i := 0; i < len(testSlice)-4; i++ {
				score = maxFloat(score, filter.tree.getScore(string(testSlice[:i+4]), modifier))
				if score > (baseConfidence + filter.toleranceModifier) {
					break
				}
			}
		} else {
			score = filter.tree.getScore(string(testSlice), modifier)
		}
		if score > (baseConfidence + filter.toleranceModifier) {
			pass = false
		}
	}
	return pass
}

func (filter *Filter) replaceSymbols (input []rune) (newSlice []rune, modifier float64){
	newSlice = []rune{}
	modifier = 0
	for _, c := range input {
		if c == ',' || c == '.' {
			modifier += .5
		} else {
			newSlice = append(newSlice, filter.wordSet.replaceSymbol(c))
		}
	}
	return
}

/* Replaces a subset of a slice with the same character. */
func replaceSlice(in []rune, char rune, start int, end int) []rune {
	for i := start; i < end; i++ {
		in[i] = char
	}
	return in
}
